### PR TITLE
Add support for checking a sentinel file for the cache

### LINKED ISSUE #1056: Add a sentinel file for the cache xrootd
In #1003 we added the option to provide a sentinel file, and it would be nice if we can do the same thing for the cache server as well.

### PR DESCRIPTION
**Pull request by haoming29**
Closes #1056

### CODE REVIEW

File: docs/parameters.yaml [hunk 1]
```diff
@@ -459,7 +459,9 @@ description: |+
       where each of these has the same effect as the corresponding "Origin.Enable*" configuration, except scoped to the
       given export. If "PublicReads" is included, "Reads" is inferred.
   - SentinelLocation: A filename under `StoragePrefix` path for Pelican to check the storage directory exists and is correctly mounted.
-      Leave it empty to skip the check. You should always choose a distinguished name for `SentinelLocation`. It should not be reused for other servers.
+      The value must be a file and contain no directory. Leave it empty to skip the check.
+
+      You should always choose a distinguished name for `SentinelLocation`. It should not be reused for other servers.
```
Comment (turetske) [inline docs/parameters.yaml:464]:
```suggestion
      You should always choose a distinct name for `SentinelLocation`. It should not be reused for other servers.
```

File: docs/parameters.yaml [hunk 2]
```diff
@@ -939,6 +941,18 @@ root_default: /run/pelican/xrootd/cache
 default: $XDG_RUNTIME_DIR/pelican/cache
 components: ["cache"]
 ---
+name: Cache.SentinelLocation
+description: |+
+  A filename under `Cache.DataLocation` path for Pelican to check the storage directory exists and is correctly mounted.
+  The value must be a file and contain no directory. Leave it empty to skip the check.
+
+  You should always choose a distinguished name for `Cache.SentinelLocation`. It should not be reused for other servers.
```
Comment (turetske) [inline docs/parameters.yaml:949]:
```suggestion
  You should always choose a distinct name for `Cache.SentinelLocation`. It should not be reused for other servers.
```

File: server_utils/origin.go [hunk 3]
```diff
@@ -492,7 +492,8 @@ from S3 service URL. In this configuration, objects can be accessed at /federati
 	return originExports, nil
 }
 
-func CheckSentinelLocation(exports []OriginExport) (ok bool, err error) {
+// Check the sentinel files from Origin.Exports
+func CheckOriginSentinelLocations(exports []OriginExport) (ok bool, err error) {
```
Comment (turetske) [inline server_utils/origin.go:496]:
```suggestion
func CheckOriginSentinelLocation(exports []OriginExport) (ok bool, err error) {
```

File: launchers/launcher.go [hunk 4]
```diff
@@ -162,7 +162,7 @@ func LaunchModules(ctx context.Context, modules config.ServerType) (servers []se
 			return
 		}
 
-		ok, err = server_utils.CheckSentinelLocation(originExports)
+		ok, err = server_utils.CheckOriginSentinelLocations(originExports)
```
Comment (turetske) [inline launchers/launcher.go:165]:
```suggestion
		ok, err = server_utils.CheckOriginSentinelLocation(originExports)
```

File: server_utils/origin_test.go [hunk 5]
```diff
@@ -437,7 +437,7 @@ func TestCheckSentinelLocation(t *testing.T) {
 		exports = append(exports, mockExportNoStn)
 		exports = append(exports, mockExportNoStn)
 
-		ok, err := CheckSentinelLocation(exports)
+		ok, err := CheckOriginSentinelLocations(exports)
```
Comment (turetske) [inline server_utils/origin_test.go:440]:
```suggestion
		ok, err := CheckOriginSentinelLocation(exports)
```

File: server_utils/origin_test.go [hunk 6]
```diff
@@ -447,7 +447,7 @@ func TestCheckSentinelLocation(t *testing.T) {
 		exports = append(exports, mockExportNoStn)
 		exports = append(exports, mockExportValidStn)
 
-		ok, err := CheckSentinelLocation(exports)
+		ok, err := CheckOriginSentinelLocations(exports)
```
Comment (turetske) [inline server_utils/origin_test.go:450]:
```suggestion
		ok, err := CheckOriginSentinelLocation(exports)
```

File: server_utils/origin_test.go [hunk 7]
```diff
@@ -458,7 +458,7 @@ func TestCheckSentinelLocation(t *testing.T) {
 		exports = append(exports, mockExportValidStn)
 		exports = append(exports, mockExportInvalidStn)
 
-		ok, err := CheckSentinelLocation(exports)
+		ok, err := CheckOriginSentinelLocations(exports)
```
Comment (turetske) [inline server_utils/origin_test.go:461]:
```suggestion
		ok, err := CheckOriginSentinelLocation(exports)
```

File: launchers/cache_serve.go [hunk 8]
```diff
@@ -51,6 +54,21 @@ func CacheServe(ctx context.Context, engine *gin.Engine, egrp *errgroup.Group, m
 		return nil, err
 	}
 
+	// Check for the sentinel file
+	if param.Cache_SentinelLocation.IsSet() {
+		sentinelPath := param.Cache_SentinelLocation.GetString()
```
Comment (turetske) [inline launchers/cache_serve.go:59]: This seems redundant. Can the server_utils function be made more general to check both the cache and origin sentinel locations?

**Review by turetske — CHANGES_REQUESTED** [review]
Can you add testing to check the cache sentinel location as well? Something similar to what's in origin_test.go?

File: server_utils/origin.go (hunk 3 above)
Comment (haoming29) [reply server_utils/origin.go:496]: The `s` was intentional as we are checking multiple locations for different prefixes

File: launchers/cache_serve.go (hunk 8 above)
Comment (haoming29) [reply launchers/cache_serve.go:59]: The first draft (not published) was to take advantage of the `server_utils` function and make them share the same core logic, but I figured the real "core" logic is just a `os.Stat` call. Given that the error message has different format for cache and origin, and that origin and cache reads from different parameters, I figured I'd be ok to have a separate function to handle them

**Review by turetske — APPROVED** [review]

**Pull request merged**